File: solutions/_622_design_circular_queue/solution.js
class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

class MyCircularQueue {
  constructor(k) {
    this.head = null;
    this.tail = null;
    this.capacity = k;
    this.size = 0;
  }

  enQueue(val) {
    if (this.isFull()) {
      return false;
    }

    const node = new Node(val);
    if (this.size === 0) {
      this.head = node;
      this.tail = node;
    } else {
      this.tail.next = node;
      this.tail = node;
    }
    this.size++;
    return true;
  }

  deQueue() {
    if (this.isEmpty()) {
      return false;
    }

    this.head = this.head.next;
    this.size--;

    return true;
  }

  Front() {
    return this.isEmpty() ? -1 : this.head.val;
  }

  Rear() {
    return this.isEmpty() ? -1 : this.tail.val;
  }

  isEmpty() {
    return this.size == 0;
  }

  isFull() {
    return this.capacity == this.size;
  }
}

module.exports = { MyCircularQueue };

/*
https://leetcode.com/problems/design-circular-queue/
*/

File: solutions/_622_design_circular_queue/solution.py
class Node:
    def __init__(self, val: int):
        self.val = val
        self.next = None


class MyCircularQueue:

    def __init__(self, k: int):
        self.head = None
        self.tail = None
        self.capacity = k
        self.size = 0

    def enQueue(self, value: int) -> bool:
        if self.isFull():
            return False

        node = Node(value)
        if self.size == 0:
            self.head = node
            self.tail = node
        else:
            self.tail.next = node
            self.tail = node

        self.size += 1

        return True

    def deQueue(self) -> bool:
        if self.isEmpty():
            return False

        self.head = self.head.next
        self.size -= 1

        return True

    def Front(self) -> int:
        return -1 if self.isEmpty() else self.head.val

    def Rear(self) -> int:
        return -1 if self.isEmpty() else self.tail.val

    def isEmpty(self) -> bool:
        return self.size == 0

    def isFull(self) -> bool:
        return self.capacity == self.size


# https://leetcode.com/problems/design-circular-queue/
